src/backend/distributed/executor/executor.c

```diff
@@ -1463,25 +1415,45 @@ ManageWorkerPool(WorkerPool *workerPool)
 		return;
```

Comment:
Not very much related with this PR.

I sometimes hit `Assert(activeConnectionCount >= idleConnectionCount);` if I randomly fail worker nodes while running SELECT queries with the flag on.

I think that's mainly because we decrement `activeConnectionCount ` on `MULTI_CONNECTION_LOST`. But, if the same connection is idle at that time, we do not decrement `idleConnectionCount`.

I think we should have probably have something like this (which fixes the issue I observe):
```
diff --git a/src/backend/distributed/executor/executor.c b/src/backend/distributed/executor/executor.c
index 66101df2..365580b1 100644
--- a/src/backend/distributed/executor/executor.c
+++ b/src/backend/distributed/executor/executor.c
@@ -1658,6 +1658,9 @@ ConnectionStateMachine(WorkerSession *session)
                                /* managed to connect, but connection was lost */
                                workerPool->activeConnectionCount--;
 
+                               if (session->currentTask == NULL)
+                                       workerPool->idleConnectionCount--;
+
                                connection->connectionState = MULTI_CONNECTION_FAILED;
                                break;
                        }
```

I can do that in another PR if you think it's not nice to have such thing here.

Reply:
Let's do it in a separate PR.